### PR TITLE
feat: add support for getElementsByClassName

### PR DESCRIPTION
**Pull request by tedconn**
I would like to add more support for more global deep traversing mechanisms:

- [x] ~~querySelector~~
- [x] ~~querySelectorAll~~
- [x] ~~getElementsByClassName~~
- [ ] getElementsByName
- [ ] getElementById
- [ ] getElementsByTagName
- [ ] getElementsByTagNameNS

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -223,7 +223,18 @@ function querySelectorAll (selector, context = document) {
   return query(selector, context, true)
 }
 
+function getElementsByClassName (classNames, context = document) {
+  const oneSpaceSeparator = classNames.trim().replace(/\s\s+/g, ' ');
```
Comment (nolanlawson) [inline src/index.js:227]: Why `/\s\s+/g` and not `/\s+/g`? We need to replace a single instance of `\t` or `\n` with a space, correct?

File: src/index.js [hunk 2]
```diff
@@ -223,7 +223,18 @@ function querySelectorAll (selector, context = document) {
...
+  const classNamesSplit = oneSpaceSeparator.split(' ');
+  let selector = "";
+  for (let i = 0; i < classNamesSplit.length; i++) {
+    selector += '.' + classNamesSplit[i];
+  }
+  return query(selector, context, true);
```
Comment (nolanlawson) [inline src/index.js:233]: Technically, `getElementsByClassName` is [supposed](https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByClassName) to return a live-updating HTMLCollection. Should we support this?

**Review by nolanlawson — CHANGES_REQUESTED** [review]
Largely LGTM.

This is increasing the scope of `kagekiri`, but the new code is still tree-shakeable and can largely piggyback off of the existing code, so I can't see a reason not to have it.

The biggest question in my mind is whether we should support live HTMLCollections or not. Some scripts may fail if they get a static array instead, but implementing a live-updating collection would be a big undertaking.

This PR would also need README changes for the new feature… Also: do you intend to complete everything in the checklist in this PR? Is this a work in progress?

File: src/index.js (hunk 2 above)
Comment (tedconn) [reply src/index.js:233]: kagekiri is stateless right? so wouldn't it be up to the consumer to implement that?

**Comment by tedconn** [comment]
> The biggest question in my mind is whether we should support live HTMLCollections or not

I mention this below/above but wouldn't that be up to whatever application is using this to introduce state? Implementing that means reimplementing an entire live DOM. kagekiri remains just a querying API? Even in Oasis we are ignoring live collections... 

> This PR would also need README changes for the new feature

Good call, I will make update the PR

> do you intend to complete everything in the checklist in this PR?

Yes, but not in this PR, just wanted to show that that's what would be done when this PR is merged and what would be left.

File: src/index.js (hunk 2 above)
Comment (nolanlawson) [reply src/index.js:233]: kagekiri is stateless, yes. I'm just calling out that if we're aiming for 1:1 compatibility with `getElementsByClassName`, then we would need to implement a live HTMLCollection.

File: src/index.js (hunk 1 above)
Comment (tedconn) [reply src/index.js:227]: When replacing with a `'  '` I think the end result is the same. `\s\s+` finds any non-single spaces where `\s+` would return single spaces.

Comment (nolanlawson) [reply src/index.js:227]: Right, but wouldn't it fail with these strings?

```js
"foo\tbar"
"foo\nbar"
```

The single `\t` or `\n` would never get replaced, so the next `.split(' ')` would not work.

Comment (tedconn) [reply src/index.js:227]: You're right, thanks! Not sure where I was going with that.

File: README.md [hunk 3]
```diff
@@ -1,7 +1,19 @@
...
+|--------------------------|-------------------|:------:|
+| `querySelector`          | Element, Document |    ✅    |
+| `querySelectorAll`       | Element, Document |    ✅    |
+| `getElementsByClassName` | Element, Document |    ✅    |
+| `getElementsByTagName`   | Element, Document |    🔜    |
+| `getElementsByTagNameNS` | Element, Document |    🔜    |
+| `getElementById`         |      Document     |    🔜    |
+| `getElementsByName`      |      Document     |    🔜    |
+| `matches`                |      Element      |    🔜    |
+
```
Comment (tedconn) [inline README.md:16]: I thought it would be weird to randomly add another API to the existing list without mentioning all the other ones. Once I finish all of the APIs we can remove the support column or format it differently. Thoughts?

File: test/test.js [hunk 4]
```diff
@@ -892,4 +896,105 @@ describe('basic test suite', function () {
...
+      withDom(classNamesLight1, context => {
+        const classNames = `	main	container
+        
+            	
+        `;
+        assertResultEqual(classNames, context.getElementsByClassName(classNames), expected, true)
+      })
+    })
+
+    it('shadow DOM - getElementsByClassName with spaces, line breaks, and tabs', () => {
```
Comment (tedconn) [inline test/test.js:990]: I added more test cases and broke them out into individual tests. In a future PR I can break this out into multiple files, eg: one per API

File: README.md (hunk 3 above)
Comment (nolanlawson) [reply README.md:16]: Makes sense to me.

File: src/index.js [hunk 5]
```diff
@@ -223,7 +223,18 @@ function querySelectorAll (selector, context = document) {
   return query(selector, context, true)
 }
 
+function getElementsByClassName (classNames, context = document) {
+  const oneSpaceSeparator = classNames.trim().replace(/\s+/g, ' ');
+  const classNamesSplit = oneSpaceSeparator.split(' ');
+  let selector = "";
+  for (let i = 0; i < classNamesSplit.length; i++) {
+    selector += '.' + classNamesSplit[i];
+  }
```
Comment (nolanlawson) [inline src/index.js:232]: Just occurred to me: instead of building up a class selector string, maybe we could use `ElementIterator` instead? I.e. iterate through the elements and check their `.classList`. The benefit is that, for folks who want to just use `getElementsByClassName`, they don't need to import all of `postcss-selector-parser` (it can be tree-shaken out).

File: types/kagekiri.d.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
 // Type definitions for kagekiri
 export function querySelector(selector: string, context?: Node): HTMLElement | null;
-export function querySelectorAll(selector: string, context?: Node): NodeList;
\ No newline at end of file
+export function querySelectorAll(selector: string, context?: Node): NodeList;
+export function getElementsByClassName(names: string, context?: Node): NodeList;
```
Comment (tedconn) [inline types/kagekiri.d.ts:4]: re `NodeList` I don't remember what we decided on but it felt weird to type it as an `HTMLCollection` when it specifically isn't.

Comment (nolanlawson) [reply types/kagekiri.d.ts:4]: It's also technically not a `NodeList` – it's just an Array of `Node`s. I mean, for most intents and purposes, it's the same, but there are subtle differences apparently: https://developer.mozilla.org/en-US/docs/Web/API/NodeList

Comment (tedconn) [reply types/kagekiri.d.ts:4]: Hmmm on the other hand, if we identify it as an HTMLCollection, at least implementors can observe and handle that accordingly.

Comment (tedconn) [reply types/kagekiri.d.ts:4]: Actually yeah then it should just be a `Node[]`

File: README.md [hunk 7]
```diff
@@ -1,7 +1,19 @@
 kagekiri [![build status](https://circleci.com/gh/salesforce/kagekiri.svg?style=svg)](https://circleci.com/gh/salesforce/kagekiri)
 ====
 
-Shadow DOM-piercing `querySelector()` / `querySelectorAll()` implementation.
+Shadow DOM-piercing query APIs
```
Comment (nolanlawson) [inline README.md:4]:
```suggestion
Shadow DOM-piercing query APIs. Supports:
```

File: README.md [hunk 8]
```diff
@@ -19,11 +31,11 @@ import { querySelector, querySelectorAll } from 'kagekiri'
 
 // query the document
 const elements = querySelectorAll('.container button')
-const element = querySelector('.container button')
+const elements = getElementsByClassName('.container button')
 
 // or a specific element
 const elements = querySelectorAll('button', otherElement)
-const element = querySelector('button', otherElement)
+const elements = getElementsByClassName('button', otherElement)
```
Comment (nolanlawson) [inline README.md:38]: I'd prefer not to change these examples. It's much less impressive to show off `getElementsByClassName` because it's more trivial to implement than a query selector engine.

The reason for class queries is just because I wanted the example selectors to be simple.

File: README.md [hunk 9]
```diff
@@ -82,7 +94,7 @@ How it works
 
 Note that it only works on open shadow DOM. Closed shadow DOM cannot be traversed.
 
-Slotted elements are considered to be children of their slots (inside the shadow DOM) rather than children of their host components. If you don't want this behavior, you can use the normal `querySelector()`/`querySelectorAll()` API.
+Slotted elements are considered to be children of their slots (inside the shadow DOM) rather than children of their host components. If you don't want this behavior, you can use the normal DOM APIs.
```
Comment (nolanlawson) [inline README.md:97]:
```suggestion
Slotted elements are considered to be children of their slots (inside the shadow DOM) rather than children of their host components. If you don't want this behavior, you can use the normal DOM APIs (e.g. `document.querySelector()` or `document.querySelectorAll()`).
```

File: src/index.js [hunk 10]
```diff
@@ -185,6 +185,22 @@ function getMatchingElements (elementIterator, ast, multiple) {
...
 
+function getMatchingElementsByClassName (elementIterator, classNames) {
+  const results = [];
+  let element
+
+  while ((element = elementIterator.next())) {
+    const elementClassList = element.classList;
+    const contains = classNames.every(function(className) {
+      return [].slice.call(elementClassList).indexOf(className) >= 0;
+    });
```
Comment (nolanlawson) [inline src/index.js:196]: Why not just `element.classList.contains()`?

File: src/index.js [hunk 11]
```diff
@@ -223,7 +239,15 @@ function querySelectorAll (selector, context = document) {
   return query(selector, context, true)
 }
 
+function getElementsByClassName (classNames, context = document) {
+  const elementIterator = new ElementIterator(context)
+  const oneSpaceSeparator = classNames.trim().replace(/\s+/g, ' ');
+  const classNamesSplit = oneSpaceSeparator.split(' ');
```
Comment (nolanlawson) [inline src/index.js:245]:
```suggestion
  const classNamesSplit = classNames.trim().split(/\s+/);
```

Comment (nolanlawson) [inline src/index.js:245]:
```suggestion
  const classNamesSplit = classNames.trim().split(/\s+/);
```

File: types/kagekiri.d.ts [hunk 12]
```diff
@@ -1,3 +1,4 @@
 // Type definitions for kagekiri
 export function querySelector(selector: string, context?: Node): HTMLElement | null;
-export function querySelectorAll(selector: string, context?: Node): NodeList;
\ No newline at end of file
+export function querySelectorAll(selector: string, context?: Node): Node[];
+export function getElementsByClassName(names: string, context?: Node): Node[];
```
Comment (nolanlawson) [inline types/kagekiri.d.ts:4]: On second thought, it seems that all of these should be `Element` rather than `HTMLElement` or `Node`. Looking at [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement), it appears that kagekiri can return Elements (e.g. HTMLElement or SVGElement – you can query SVGs just like HTML), but not generic Nodes (e.g. comments, text).

File: src/index.js [hunk 13]
```diff
@@ -185,6 +185,22 @@ function getMatchingElements (elementIterator, ast, multiple) {
...
 }
 
+function getMatchingElementsByClassName (elementIterator, classNames) {
+  const results = [];
+  let element
+
+  while ((element = elementIterator.next())) {
+    const elementClassList = element.classList;
+    const contains = classNames.every(function(className) {
+      return [].slice.call(elementClassList).indexOf(className) >= 0;
```
Comment (tedconn) [inline src/index.js:195]:
```suggestion
      return elementClassList.contains(className);
```

File: src/index.js (hunk 10 above)
Comment (tedconn) [reply src/index.js:196]: Great catch, I confused `contains` with `includes`

File: types/kagekiri.d.ts (hunk 12 above)
Comment (tedconn) [reply types/kagekiri.d.ts:4]: Yeah the documentation is confusing. Apparently `querySelectorAll` [returns](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll#Syntax):

> **A non-live NodeList containing one Element object for each element that matches** at least one of the specified selectors or an empty NodeList in case of no matches.

A NodeList containing Elements? Also, I'm not aware of any way to query non-element nodes (text or comment) using `querySelectorAll`.

What do you want to do? My vote would be for `Node[]` -- IMO it aligns nicely with the spec and what we actually return, but up to you.

File: README.md (hunk 8 above)
Comment (tedconn) [reply README.md:38]: Sure thing! Will revert.

File: types/kagekiri.d.ts (hunk 12 above)
Comment (nolanlawson) [reply types/kagekiri.d.ts:4]: Despite the name, it seems like `NodeList` actually contains `Element`s, not `Node`s.

I tested, and kagekiri can definitely return `SVGElement` objects (e.g. a `<g>` tag). However, I don't see how it can return generic [Node](https://developer.mozilla.org/en-US/docs/Web/API/Node)s, which include:

> Attr, CharacterData (which Text, Comment, and CDATASection are all based on), ProcessingInstruction, DocumentType, Notation, Entity, and EntityReference

So my vote would be for `Element[]`, since I can't think of any `Node` but non-`Element` that kagekiri can return.

File: README.md [hunk 14]
```diff
@@ -19,7 +31,7 @@ import { querySelector, querySelectorAll } from 'kagekiri'
 
 // query the document
 const elements = querySelectorAll('.container button')
-const element = querySelector('.container button')
+const element = querySelector('button')
```
Comment (nolanlawson) [inline README.md:34]: Can we undo this change, please?

File: types/kagekiri.d.ts [hunk 15]
```diff
@@ -1,3 +1,4 @@
 // Type definitions for kagekiri
 export function querySelector(selector: string, context?: Node): HTMLElement | null;
```
Comment (nolanlawson) [inline types/kagekiri.d.ts:2]: `querySelector` should return an `Element`, not an `HTMLElement`. It's possible to query an SVG here and get an `SVGElement`.

**Review by nolanlawson — APPROVED** [review]

**Comment by nolanlawson** [comment]
@tedconn Just noticed – did you mean to use `classNamesLight2` and `classNamesShadow2` somewhere in your tests? They're unused.

**Comment by tedconn** [comment]
> @tedconn Just noticed – did you mean to use `classNamesLight2` and `classNamesShadow2` somewhere in your tests? They're unused.

good catch, I removed them in #28
